Datacenter Picker: Allow user to select datacenter when going Atomic

danielbachhuber opened the pull request:
See https://github.com/Automattic/dotcom-forge/issues/1237

Some design conversation in p1669037882422489-slack-C041RHH38NQ
And more conversation in p1669156646360569-slack-C7YPW6K40
And a design request in p9Jlb4-5Xy-p2

## Proposed Changes

When a user takes their site Atomic from 'Hosting Configuration', they'll see a message about their site being automatically placed in the best datacenter:

<img width="842" alt="image" src="https://user-images.githubusercontent.com/36432/203596511-d744aef0-e815-4a46-826e-5955390bdb28.png">

If they click on 'Choose a datacenter instead', they're presented UI to pick their datacenter:

<img width="833" alt="image" src="https://user-images.githubusercontent.com/36432/203626960-23acbe0f-90fb-4482-85c5-847d980dee46.png">

If they don't have any eligibility warnings, they'll see this:

<img width="821" alt="image" src="https://user-images.githubusercontent.com/36432/203596908-2654f3bb-8267-4ca8-8224-5ec0dac732ce.png">

They won't be presented with the datapicker UI if they go Atomic outside of 'Hosting Configuration':

<img width="1171" alt="image" src="https://user-images.githubusercontent.com/36432/203597358-6fa73ba2-d2ea-489b-8f30-f1f730f0659e.png">

## Testing Instructions

1. First, sandbox `public-api.wordpress.com` and sandbox D93223-code.
2. Create a new WordPress.com site with a business plan.
3. Attempt to install a free plugin and verify the `<EligibilityWarnings>` component displays without the Datacenter Picker UI.
4. Navigate to 'Hosting Configuration' and click 'Activate'.
5. Verify the `<EligibilityWarnings>` component displays with the Datacenter Picker UI.
6. Verify the Datacenter Picker UI is usable, accessible, and looks good in the browsers we support.
7. If you want to recreate the 'no eligibility warnings' state, map a domain to your site and then go back to 'Hosting Configuration' -> 'Activate'.
8. Pick a custom datacenter, click 'Continue', and verify the site is created in the expected datacenter.

danielbachhuber committed:
First pass at UI for datacenter picker

danielbachhuber committed:
Remove unnecessary `map()`

danielbachhuber committed:
Only show in `development` and `wpcalypso` envirnoments for now

danielbachhuber committed:
Refactor to use `<FoldableCard>` inside `<EligibilityWarnings>`

danielbachhuber committed:
Merge branch 'trunk' into improve/datacenter-picker-transfer-ui

danielbachhuber committed:
Only show the datacenter picker on 'Activate Hosting Features'

danielbachhuber committed:
First pass at passing `geo_affinity` to the transfer process

danielbachhuber committed:
Merge branch 'trunk' into improve/datacenter-picker-transfer-ui

danielbachhuber committed:
Refactor datacenter picker styling

danielbachhuber committed:
More datacenter picker UI tweaking

danielbachhuber committed:
More UI tweaks

danielbachhuber committed:
Clean up `onProceed` types

danielbachhuber committed:
Language tweaking

danielbachhuber committed:
Merge branch 'trunk' into improve/datacenter-picker-transfer-ui

p-jackson commented on client/blocks/eligibility-warnings/warning-list.tsx:
```diff
@@ -1,6 +1,5 @@
 import { Card } from '@automattic/components';
 import { localize, LocalizeProps, translate } from 'i18n-calypso';
-import { map } from 'lodash';
```
😎

p-jackson commented on client/my-sites/hosting/hosting-activate.js:
```diff
@@ -11,8 +11,8 @@ import { getSelectedSiteId, getSelectedSiteSlug } from 'calypso/state/ui/selecto
 
-	const transferInitiate = () => {
-		initiateTransfer( siteId, null, null );
+	const transferInitiate = ( { geo_affinity = null } ) => {
```
Everywhere else `geo_affinity` is defaulted it uses `''` to ensure the value is always a string. Personally I like using `null` to represent empty 😄 but why break the convention for this case?

danielbachhuber replied:
> Everywhere else `geo_affinity` is defaulted it uses `''` to ensure the value is always a string. Personally I like using `null` to represent empty 😄 but why break the convention for this case?

Good point. Updated with 3ffa6e64b4a257210f64c37dfda57d00fe99a559

p-jackson commented on client/state/themes/actions/theme-transfer.js:
```diff
@@ -12,7 +12,7 @@ import {
 import 'calypso/state/themes/init';
 
-function initiateTransfer( siteId, plugin, theme, onProgress ) {
+function initiateTransfer( siteId, plugin, theme, geoAffinity, onProgress ) {
```
🌶️ Living dangerously and adding new params in the middle of the param list! 😥 

I did a search and it looks like you've caught all the cases thought 👍

danielbachhuber replied:
The function isn't exported, fortunately!

p-jackson commented on client/blocks/eligibility-warnings/datacenter-picker.tsx:
```diff
@@ -0,0 +1,155 @@
+import { localizeUrl } from '@automattic/i18n-utils';
+import styled from '@emotion/styled';
```
We're not using Emotion anywhere else in `client/blocks/eligibility-warnings` so I'd be reluctant to add it for this component without a compelling reason. My guess this is what's adding the few extra KBs to the async bundles.

danielbachhuber replied:
> We're not using Emotion anywhere else in `client/blocks/eligibility-warnings` so I'd be reluctant to add it for this component without a compelling reason.

My understanding is that we want to use CSS objects for all new components, and when we have an opportunity to refactor existing components.

p-jackson replied:
Interesting! That's not my impression at all. Seemed like there have been a number of conversations like this p1666216809948199-slack-C02A41GCS where folks have been questioning its usefulness.

p-jackson commented on client/blocks/eligibility-warnings/datacenter-picker.tsx:
```diff
@@ -0,0 +1,155 @@
+						{ translate( 'Your site will be automatically placed in the best datacenter.' ) }
+					</span>
+					&nbsp;
+					<Button isTertiary={ true } onClick={ () => setIsFormShowing( ! isFormShowing ) }>
```
Continuing the "which Button should I pick" conversation: another option would be to use the button from `@automattic/components` because it has a `borderless` prop. This removes the horizontal padding and makes it more like a text link.

You might have to fight with the old `inline-block` vertical-alignment issues though.

One button isn't "better" than the other though. I do like the idea of using WP components wherever possible since they're the lowest-common denominator and most portable.

danielbachhuber replied:
> Continuing the "which Button should I pick" conversation: another option would be to use the button from `@automattic/components` because it has a `borderless` prop. This removes the horizontal padding and makes it more like a text link.

@p-jackson I considered that! The `@automattic/components` `Button` doesn't have `isTertiary`, which is why I used `@wordpress/components`.

p-jackson replied:
[This comment explains it](https://github.com/Automattic/wp-calypso/pull/70237#issuecomment-1329154498) thanks 😅

p-jackson commented on client/blocks/eligibility-warnings/datacenter-picker.tsx:
```diff
@@ -0,0 +1,155 @@
+	display: 'block',
+	marginBottom: '10px',
+	width: '100%',
+	maxWidth: '541px',
```
This max width works for the usual layout, but not when the screen is narrower. Nothing breaks though, and you probably don't want to come up with multiple SVGs for all the different layouts.

p-jackson replied:
You should also add `cursor: pointer` to `<AutomaticFormLabel>` so that it matches `<FormRadiosBar>`

danielbachhuber replied:
> You should also add `cursor: pointer` to `<AutomaticFormLabel>` so that it matches `<FormRadiosBar>`

ff53dc2e577e2c7496c7cddf4f8016e480728d05

p-jackson reviewed (commented):
The approach looks good to me 👍 

There's an issue with the radio buttons in that the usual left-right keyboard shortcuts aren't working. This is because the buttons aren't being tied together by a common `name` HTML attribute.

If you add a `name: 'datacenter'` attribute to all of the elements in `DatacenterOptions`, and then add a `name="datacenter"` prop to the automatic radio button, then it'll fix up the semantics (yay a11y) and the keyboard shortcuts will automatically start working.

danielbachhuber committed:
Merge branch 'trunk' into improve/datacenter-picker-transfer-ui

danielbachhuber committed:
Mirror cursor effect of other labels

danielbachhuber committed:
Default to an empty string for type consistency

danielbachhuber committed:
Tie all of the inputs together with `name="geo_affinity"`

danielbachhuber commented:
> There's an issue with the radio buttons in that the usual left-right keyboard shortcuts aren't working. This is because the buttons aren't being tied together by a common `name` HTML attribute.
> 
> If you add a `name: 'datacenter'` attribute to all of the elements in `DatacenterOptions`, and then add a `name="datacenter"` prop to the automatic radio button, then it'll fix up the semantics (yay a11y) and the keyboard shortcuts will automatically start working.

@p-jackson Interesting! This actually worked in my testing on Chrome after I added `role="radiogroup"` to the container.

In any case, I've added `name="geo_affinity"` to the radio buttons: 8909b8e07512ec0762163e7ffbe27581f00ff474

danielbachhuber committed:
Track interactions with the datacenter picker

wojtekn reviewed (commented):
@danielbachhuber feature-wise works well for me. I left one question related to server-side testing under the diff.

I see one thing related to the UI - the "Choose a datacenter instead" and "Cancel" look like a link, but when we hover them, they look like buttons. Is it intended? Would it make sense to style them just like a like, like other secondary actions, e.g., "Contact support" or "Learn more"?

danielbachhuber committed:
Track when the component is displayed, too

danielbachhuber commented:
> I see one thing related to the UI - the "Choose a datacenter instead" and "Cancel" look like a link, but when we hover them, they look like buttons. Is it intended? Would it make sense to style them just like a like, like other secondary actions, e.g., "Contact support" or "Learn more"?

@wojtekn AAAAAAAAAAAH!!!!!! (screams in terror)

I originally had it as a link (p1669037882422489-slack-C041RHH38NQ). Marko suggested that "textual links create the expectation it will navigate to another page" (p1669038167582919/1669037882.422489-slack-C041RHH38NQ). Saxon suggested the "tertiary button style" (p1669157399844259/1669037882.422489-slack-C041RHH38NQ). I wasn't able to find a tertiary button style in `@automattic/components`, so I went with `@wordpress/components`.

And here we are!

In all seriousness, I don't have a strong preference but I'm going to stick with what I have.

wojtekn commented:
> In all seriousness, I don't have a strong preference but I'm going to stick with what I have for now.

Oh, thanks for providing more background on that. It was something that caught my eye, but I don't have strong arguments against that 👍
